### ISSUE TITLE
[ISSUE] `service.type` field could be a string or an array of strings according to the did core specification

### ISSUE THREAD

**Issue by PatStLouis** [Indicio-tech/pydid#81]
Raising this issue here based on [this aca-py issue](https://github.com/hyperledger/aries-cloudagent-python/issues/2677#issue-2044982863). According to the did core spec, the `service.type` value of a did document could be a string or an array of strings. This prevents applications leveraging pyDid such as aca-py from being interoperable with other platforms using a `service.type` array. In a scenario where we would want to verify a VC issued by an issuer who has a `service.type` array in it's did document, said application will not be able to fetch the verification key since the parsing of the did document will fail.

I would like to understand more the scope of changing this behavior.

Would changing [this line](https://github.com/Indicio-tech/pydid/blob/9fe696f93b91a1654ecc2ad8174fba96222dd58d/pydid/service.py#L20) to `type: Union[str, List[str]]` help solve this issue @dbluhm ?

**Comment by dbluhm** [comment]
@PatStLouis that would be the minimum required to get the core service type to handle lists of types. This raises the question of how we should narrow a service to a DIDComm service, which is a key function that downstream users depend on. We could leave it at expecting DIDComm services to not be a list of strings and postpone making decisions about what the behavior of a service with type `DIDCommMessaging` and others in the same list mean.

**Comment by PatStLouis** [comment]
The most important is to enable pydid to **resolve/read** did documents correctly. I'm not too concerned about the creation of did documents/services with pydid at the moment.

I have yet to see a service type list containing more than 1 item so for the cases where one would expect a string, maybe a `serviceType = serviceType[0] if isinstance(serviceType, list) else serviceType` one liner logic could solve this? The didcomm spec could also specify that the service type MUST be a string when registering the service.

To visualize and make sure we are on the same page, here's an example where both scenarios are in the same did document:
```
{
  "services": [
    {
      "id": "did:web:example.com#traceability-api",
      "type": [
        "TraceabilityApi"
      ],
      "service_endpoint": "https://traceability.example.com"
    },
    {
      "id": "did:web:example.com#didcomm-messaging",
      "type": "DIDCommMessaging",
      "service_endpoint": "https://didcomm.example.com"
    }
  ]
}
```

A did document containing this service section wouldn't be resolvable with the  `/resolver/resolve` in aca-py currently, even if it's a valid document. It should at the very least be able to resolve it. This also prevents the new `/vc/ldp/verify` routes from resolving a verification method in a document where there is a service type list, which is unrelated to the task IMO.

And to be clear I'm not debating if having an array is a good this or not, I'm just looking to adhere to the didcore spec and improve interoperability.

**Comment by dbluhm** [comment]
@PatStLouis yes, I understand; my comments are likewise focused on the correct resolution and parsing of DID Documents. For now, let's change the base service class to accept the union of strings or a list of strings as you suggest and we will ignore the case where a DIDComm service is potentially declared with the type as a list. If you have some time, a PR with this change and some test cases are welcome.

**Comment by PatStLouis** [comment]
Sounds good, I'll submit something this week for review

**Comment by PatStLouis** [comment]
PR is in #82

**Comment by dbluhm** [comment]
These has been merged and released in 0.4.3. I will create a separate issue to track the question of whether the DIDComm service type should permit a list of strings or not.
